### PR TITLE
RFE: TestRuleAddAndLoad: fix for ppc

### PR DESCRIPTION
**Pull request by kolyshkin**
As reported in [1], the test fails on ppc64le since getpid(2) is
not supposed to ever return an error, and somehow glibc relies on
that assumption, and returns a positive value of error set by
the seccomp rule from its getpid(2) wrapper. This makes the test fail.

While seccomp_rule_add(3) man page was amended to note that behavior
in [2], the test case here was never fixed so it always fails on ppc.

Fix it by replacing getpid(2) with close(2). Make some other cosmetic
changes while at it.

The test case was tested to fail (when the corresponding FilterAddRule
call is commented out):

> seccomp_test.go:647: Syscall close: want no space left on device, got bad file descriptor

[1] https://github.com/seccomp/libseccomp-golang/issues/61
[2] https://github.com/seccomp/libseccomp/pull/333

Fixes: #61.

### CODE REVIEW

**Comment by kolyshkin** [comment]
Alas we don't have a (simple) way to test this on PPC (yet). In theory, we could use qemu in GHA CI -- this is something that I might take a look at later. For now, if someone can run the test before and after this PR, that'd be awesome.

**Comment by kolyshkin** [comment]
Found by looking into Debian package, which has the following additional patch:
```diff
From: Shengjing Zhu <zhsj@debian.org>
Date: Tue, 14 Jun 2022 03:30:34 +0800
Subject: Skip TestRuleAddAndLoad on ppc64le (Closes: #976943)

---
 seccomp_test.go | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/seccomp_test.go b/seccomp_test.go
index c65dbf0..408b673 100644
--- a/seccomp_test.go
+++ b/seccomp_test.go
@@ -6,6 +6,7 @@ import (
 	"fmt"
 	"os"
 	"os/exec"
+	"runtime"
 	"strings"
 	"syscall"
 	"testing"
@@ -579,6 +580,9 @@ func subprocessAddRuleErrors(t *testing.T) {
 }
 
 func TestRuleAddAndLoad(t *testing.T) {
+	if runtime.GOARCH == "ppc64le" {
+		t.Skip("https://github.com/seccomp/libseccomp-golang/issues/61")
+	}
 	execInSubprocess(t, subprocessRuleAddAndLoad)
 }
```

The patch file name (0001-Skip-TestRuleAddAndLoad-on-ppc64le-Closes-976943.patch) apparently refers to a debian bug: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=976943

**Comment by zhsj** [comment]
```
=== RUN   TestRuleAddAndLoad
    seccomp_test.go:39: === RUN   TestRuleAddAndLoad
        === RUN   TestRuleAddAndLoad/subprocess
        --- PASS: TestRuleAddAndLoad (0.00s)
            --- PASS: TestRuleAddAndLoad/subprocess (0.00s)
        PASS
        testing: can't write /tmp/go-build1265467549/b001/testlog.txt: close /tmp/go-build1265467549/b001/testlog.txt: no space left on device
    seccomp_test.go:41: exit status 2
--- FAIL: TestRuleAddAndLoad (0.00s)
```

Still failing.

**Comment by zhsj** [comment]
However running the single test will pass

```
$ go test -v -run TestRuleAddAndLoad
=== RUN   TestRuleAddAndLoad
    seccomp_test.go:39: === RUN   TestRuleAddAndLoad
        === RUN   TestRuleAddAndLoad/subprocess
        --- PASS: TestRuleAddAndLoad (0.00s)
            --- PASS: TestRuleAddAndLoad/subprocess (0.00s)
        PASS
--- PASS: TestRuleAddAndLoad (0.01s)
PASS
ok      github.com/seccomp/libseccomp-golang    0.011s
```

**Comment by kolyshkin** [comment]
Yeah, apparently `close()` is also a bad choice, since it's used by the test runnner/logger itself. Need to find some other simple syscall which we can block without any bad consequences.

**Comment by kolyshkin** [comment]
Changed `close()` to `listen()`; can you please test it @zhsj ?

**Comment by zhsj** [comment]
All pass now

```
(sid_ppc64el-dchroot)zhsj@plummer:~/libseccomp-golang$ go test -v .
=== RUN   TestExpectedSeccompVersion
    seccomp_test.go:39: === RUN   TestExpectedSeccompVersion
        === RUN   TestExpectedSeccompVersion/subprocess
            seccomp_test.go:55: _EXPECTED_LIBSECCOMP_VERSION not set
        --- PASS: TestExpectedSeccompVersion (0.00s)
            --- SKIP: TestExpectedSeccompVersion/subprocess (0.00s)
        PASS
--- PASS: TestExpectedSeccompVersion (0.01s)
=== RUN   TestGetAPILevel
    seccomp_test.go:39: === RUN   TestGetAPILevel
        === RUN   TestGetAPILevel/subprocess
            seccomp_test.go:90: Got API level of 4
        --- PASS: TestGetAPILevel (0.00s)
            --- PASS: TestGetAPILevel/subprocess (0.00s)
        PASS
--- PASS: TestGetAPILevel (0.01s)
=== RUN   TestSetAPILevel
    seccomp_test.go:39: === RUN   TestSetAPILevel
        === RUN   TestSetAPILevel/subprocess
        --- PASS: TestSetAPILevel (0.00s)
            --- PASS: TestSetAPILevel/subprocess (0.00s)
        PASS
--- PASS: TestSetAPILevel (0.00s)
=== RUN   TestActionSetReturnCode
--- PASS: TestActionSetReturnCode (0.00s)
=== RUN   TestSyscallGetName
    seccomp_test.go:140: Got name of syscall 0x1 on native arch as exit
--- PASS: TestSyscallGetName (0.00s)
=== RUN   TestSyscallGetNameByArch
--- PASS: TestSyscallGetNameByArch (0.00s)
=== RUN   TestGetSyscallFromName
    seccomp_test.go:185: Got syscall number of write on native arch as 4
--- PASS: TestGetSyscallFromName (0.00s)
=== RUN   TestGetSyscallFromNameByArch
    seccomp_test.go:203: Got syscall number of write on AMD64 as 1
--- PASS: TestGetSyscallFromNameByArch (0.00s)
=== RUN   TestMakeCondition
--- PASS: TestMakeCondition (0.00s)
=== RUN   TestGetNativeArch
    seccomp_test.go:266: Got native arch of system as ppc64le
--- PASS: TestGetNativeArch (0.00s)
=== RUN   TestFilterCreateRelease
--- PASS: TestFilterCreateRelease (0.00s)
=== RUN   TestFilterReset
--- PASS: TestFilterReset (0.00s)
=== RUN   TestFilterArchFunctions
--- PASS: TestFilterArchFunctions (0.00s)
=== RUN   TestFilterAttributeGettersAndSetters
--- PASS: TestFilterAttributeGettersAndSetters (0.00s)
=== RUN   TestMergeFilters
--- PASS: TestMergeFilters (0.00s)
=== RUN   TestAddRuleErrors
    seccomp_test.go:39: === RUN   TestAddRuleErrors
        === RUN   TestAddRuleErrors/subprocess
        --- PASS: TestAddRuleErrors (0.00s)
            --- PASS: TestAddRuleErrors/subprocess (0.00s)
        PASS
--- PASS: TestAddRuleErrors (0.00s)
=== RUN   TestRuleAddAndLoad
    seccomp_test.go:39: === RUN   TestRuleAddAndLoad
        === RUN   TestRuleAddAndLoad/subprocess
        --- PASS: TestRuleAddAndLoad (0.00s)
            --- PASS: TestRuleAddAndLoad/subprocess (0.00s)
        PASS
--- PASS: TestRuleAddAndLoad (0.00s)
=== RUN   TestLogAct
    seccomp_test.go:39: === RUN   TestLogAct
        === RUN   TestLogAct/subprocess
        --- PASS: TestLogAct (0.00s)
            --- PASS: TestLogAct/subprocess (0.00s)
        PASS
--- PASS: TestLogAct (0.00s)
=== RUN   TestCreateActKillThreadFilter
    seccomp_test.go:39: === RUN   TestCreateActKillThreadFilter
        === RUN   TestCreateActKillThreadFilter/subprocess
        --- PASS: TestCreateActKillThreadFilter (0.00s)
            --- PASS: TestCreateActKillThreadFilter/subprocess (0.00s)
        PASS
--- PASS: TestCreateActKillThreadFilter (0.00s)
=== RUN   TestCreateActKillProcessFilter
    seccomp_test.go:39: === RUN   TestCreateActKillProcessFilter
        === RUN   TestCreateActKillProcessFilter/subprocess
        --- PASS: TestCreateActKillProcessFilter (0.00s)
            --- PASS: TestCreateActKillProcessFilter/subprocess (0.00s)
        PASS
--- PASS: TestCreateActKillProcessFilter (0.00s)
=== RUN   TestNotif
    seccomp_test.go:39: === RUN   TestNotif
        === RUN   TestNotif/subprocess
            seccomp_test.go:804: seccomp notification requires libseccomp >= 2.5.0 and API level >= 6 (current version: 2.5.4, API level: 4)
        --- PASS: TestNotif (0.00s)
            --- SKIP: TestNotif/subprocess (0.00s)
        PASS
--- PASS: TestNotif (0.00s)
=== RUN   TestNotifUnsupported
    seccomp_test.go:39: === RUN   TestNotifUnsupported
        === RUN   TestNotifUnsupported/subprocess
        --- PASS: TestNotifUnsupported (0.00s)
            --- PASS: TestNotifUnsupported/subprocess (0.00s)
        PASS
--- PASS: TestNotifUnsupported (0.00s)
=== RUN   TestCheckVersion
    seccomp_ver_test.go:25: verNew requires libseccomp >= 100.99.7 (current version: 2.5.4)
    seccomp_ver_test.go:25: verMajor+1 requires libseccomp >= 3.0.0 (current version: 2.5.4)
    seccomp_ver_test.go:25: verMinor+1 requires libseccomp >= 2.6.0 (current version: 2.5.4)
    seccomp_ver_test.go:25: verMicro+1 requires libseccomp >= 2.5.5 (current version: 2.5.4)
    seccomp_ver_test.go:25: <nil>
    seccomp_ver_test.go:25: <nil>
--- PASS: TestCheckVersion (0.00s)
=== RUN   TestCheckAPI
    seccomp_ver_test.go:59: apiHigh requires libseccomp >= 0.0.0 and API level >= 99 (current version: 2.5.4, API level: 4)
    seccomp_ver_test.go:59: api+1 requires libseccomp >= 0.0.0 and API level >= 5 (current version: 2.5.4, API level: 4)
    seccomp_ver_test.go:59: <nil>
    seccomp_ver_test.go:59: <nil>
    seccomp_ver_test.go:59: <nil>
    seccomp_ver_test.go:59: verHigh requires libseccomp >= 99.0.0 (current version: 2.5.4)
--- PASS: TestCheckAPI (0.00s)
PASS
ok      github.com/seccomp/libseccomp-golang    0.051s
```

File: seccomp_test.go [hunk 1]
```diff
@@ -590,9 +590,14 @@ func subprocessRuleAddAndLoad(t *testing.T) {
 	}
 	defer filter1.Release()
 
-	call, err := GetSyscallFromName("getpid")
+	const expErr = 28 // ENOSPC, but can be anything not usually returned by listen(2).
```
Comment (rata) [inline seccomp_test.go:593]: nit: can't we use unix.ENOSPC instead? (or syscall.ENOSPC for that matter, if we are using the deprecated syscall module here, as it seems)

**Review by rata — APPROVED** [review]
LGTM

**Review by pcmoore — APPROVED** [review]
Looks good to me, I'll merge in just a second.

**Comment by pcmoore** [comment]
Merged via 9006a281d6713ba604f36f9b5caeac23a16bbd55, thanks everyone!

File: seccomp_test.go (hunk 1 above)
Comment (kolyshkin) [reply seccomp_test.go:593]: You are right, not sure why I overlooked it.

Addressed as a follow up here: https://github.com/seccomp/libseccomp-golang/pull/102

**Pull request closed without merging**